### ISSUE TITLE
Implement cachevote instead gc vote

### ISSUE THREAD

**Issue by Birdy2404** [cgeo/cgeo#9664]
Hello,
as gcvote is not maintained anymore, https://cachevote.schedler.xyz/ seems to be the successor of it.
An import of gcvotes into cachevote is possible.
Would be great to see it implemented in c:geo.
Thanks in advance and best regards from South Baden
Andreas

**Comment by hakuchi** [comment]
Is there a way to extract the full database from GcVote, because there are over 8.000.000 votes?
cgeo was able to display votes from GcVote without username and password before I added my logins.

**Comment by SammysHP** [comment]
- Where is the API documentation? Was not able to find it.
- Can votes be queried without login?
- It will be confusing to show multiple voting systems in a cache listing. Either we have to merge the votes somehow or they must all be imported into CacheVote
- Can the server handle so many requests? c:geo will request the vote for every cache opened or downloaded. There might me several hundred requests per second in high load scenarios.
- This (or something else) will probably replace the voting functionality in c:geo once it is mature. Or we have to implement multiple providers (in the UI).

**Comment by Lineflyer** [comment]
@Birdy2404 
Are you involved in CacheVote operation/development?

I read there, that it shall be the successor of GCVote. The GCVote service seems to be running rather normal at this moment and last time I had contact with BonnerGuido he was able to revive the server. It would be interesting to know if and when a shutdown of GCVote will be done.

**Comment by Lineflyer** [comment]
I contacted the author of CacheVote in parallel and linked him to this thread.

**Comment by Birdy2404** [comment]
> I contacted the author of CacheVote in parallel and linked him to this thread.

That's what I thought to do. Thank you very much.

**Comment by dev-Fabi** [comment]
Hello, I'm the developer of CacheVote.

I was not aware that somebody already opened an Issue for me. Wanted to wait till I have finished the API docs...

After consulting BonnerGuido (developer of gcVote) , I started programming a new edition of gcVote called CacheVote. (so BonnerGuido is aware of that and is supporting me)
We haven't talked about if he will shutdown gcVote when CacheVote has reached a "good user base".

> Is there a way to extract the full database from GcVote, because there are over 8.000.000 votes?

Not really...
You probably could extract the average rating and count of ratings for each cache but there is no possibility to get information about the user which made the ratings (which would then end in multiple dublicate ratings).
The developer of gcVote promised never to pass the reviews on. Because of that every user has to import the ratings by himself (CacheVote handels this if the user supplies the credentials)

> Where is the API documentation? Was not able to find it.

Documentation for the API is work in progress (its a simple json REST API)

> Can votes be queried without login?

Yes it is possible (as it was in gcVote)

> It will be confusing to show multiple voting systems in a cache listing. Either we have to merge the votes somehow or they must all be imported into CacheVote

I agree showing multiple systems is confusing. I thought about deciding for each cache individually based on which one provides more ratings (which would end in two request for each cache wich I think is not ideal) or simply decide based on which account inforamtions are supplied (and fallback to a default)

> Can the server handle so many requests? c:geo will request the vote for every cache opened or downloaded. There might me several hundred requests per second in high load scenarios.

As CacheVote was just released a week ago I'm not sure (will do a stress test) but that shouldn't be a problem. The server could be upgraded if needed.

> Is it even possible to vote with gcVote currently?

As long as you already have an account and you know your password it is possible, but it is not maintained anymore

**Comment by Lineflyer** [comment]
> Hello, I'm the developer of CacheVote.

Wow...that was a fast reply :)

> The developer of gcVote promised never to pass the reviews on. Because of that every user has to import the ratings by himself (CacheVote handels this if the user supplies the credentials)

That might even be the better approach and a fresh start (with optional take over).

> I agree showing multiple systems is confusing. I thought about deciding for each cache individually based on which one provides more ratings (which would end in two request for each cache wich I think is not ideal) or simply decide based on which account inforamtions are supplied (and fallback to a default)

I would suggest to simply change to CacheVote in c:geo without any double implementation.
For users requesting votes maybe you could think about to fallback on server-side and request votes on GCVote if no/not enough votes are available in your own DB. (Meaning CacheVote with a flexible proxy towards GCVote for an interim period). Would that be an option?

For providing votes I think BonnerGuido should inform registered users (if this is technically possible at all) to change to CacheVote and after some period block voting access (providing votes).

Last but not least my personal wishlist: :)
- Would you think about also supporting other platforms than geocaching.com. As its called CacheVote instead of GCVote this would only be logical ;) and highly appreciated from my side
- Did you think about offering a more fine grained scale for ratings than just 5 levels?

**Comment by Lineflyer** [comment]
@dev-Fabi 
I just registered, confirmed mail and took a first look at the browser plugin. Nice.

I also tried to take over GCVotes, but nothing happens after entering credentials and clicking on "Jetzt importieren".  Don't see any action performed or network traffic.

**Comment by Lineflyer** [comment]
> I also tried to take over GCVotes, but nothing happens after entering credentials and clicking on "Jetzt importieren". Don't see any action performed or network traffic.

As I just returned to geocaching.com I do see the CacheVote for caches I voted on GCVote. Seemed it was only a bit unresponsive but it did its job. Now the setting is gone as expected ("Du hast deine gcVote Bewertungen bereits importiert.").

**Comment by dev-Fabi** [comment]
@Lineflyer 
> For users requesting votes maybe you could think about to fallback on server-side and request votes on GCVote if no/not enough votes are available in your own DB. Would that be an option?

Haven't thought about that... definitely an option.

> Would you think about also supporting other platforms than geocaching.com. As its called CacheVote instead of GCVote this would only be logical ;) and highly appreciated from my side

Some other cachers already asked me that. I have not really thought about that but I can imagine that other platforms will also be supported in the future.

> Did you think about offering a more fine grained scale for ratings than just 5 levels? Edit: I have just seen it does not support half stars, so its even worse than on GCVote. For me personally only 5 levels of cache quality is really difficult sometimes.

I have racked my brains over this for some time and finally decided for only 5 stars due to siplicity (for the user) and because almost all rating base on 1 to 5 stars. I know sometimes it is not easy to decide but also often more options doesn't make it easier...

> I also tried to take over GCVotes, but nothing happens after entering credentials and clicking on "Jetzt importieren". Don't see any action performed or network traffic.
As I just returned to geocaching.com I do see the CacheVote for caches I voted on GCVote. Seemed it was only a bit unresponsive but it did its job. Now the setting is gone as expected ("Du hast deine gcVote Bewertungen bereits importiert.").

It's all done serverside with one request and a success response at the end - should maybe be changed as the imports can take quite some time.

**Comment by Lineflyer** [comment]
> It's all done serverside with one request and a success response at the end - should maybe be changed as the imports can take quite some time.

I thought so..but it read "just takes a few seconds". Seemed to took longer to take over my approx. 6000 votes :)

> I have racked my brains over this for some time and finally decided for only 5 stars due to siplicity (for the user) and because almost all rating base on 1 to 5 stars. I know sometimes it is not easy to decide but also often more options doesn't make it easier...

Yeah, just thought about it over and over...and you are probably right. Still I would miss the half stars which I use all the time when I am undecided between two levels. That would also be possible while keeping the importability from GCvote.

**Comment by SammysHP** [comment]
Fewer options force you to make a decision and increases the contrast. That's why so many services use five star user rating, but half star result.

**Comment by Lineflyer** [comment]
@dev-Fabi 

> > For users requesting votes maybe you could think about to fallback on server-side and request votes on GCVote if no/not enough votes are available in your own DB. Would that be an option?
>
> Haven't thought about that... definitely an option.

AFAICS you have implemented this meanwhile?! 
Because I just have seen a cache page where a rating with 3 votes was present and after I provided my own vote the rating was reset to 1 vote (=my vote). Is that due to that implementation or was it a bug?

Any news on an API btw?

**Comment by dev-Fabi** [comment]
@Lineflyer 

Yes this feature is already implemented. But there are still some bugs.


> Any news on an API btw?

Not really. As I have some exams at the moment, there is not much time left for the project. I think I will have the API ready for "public access" by early/mid February.

**Comment by lukeIam** [comment]
> Not really...
> You probably could extract the average rating and count of ratings for each cache but there is no possibility to get information about the user which made the ratings (which would then end in multiple dublicate ratings).
> The developer of gcVote promised never to pass the reviews on. Because of that every user has to import the ratings by himself (CacheVote handels this if the user supplies the credentials)

Hi @dev-Fabi,

just a small note from an ex-developer of GCVote:

The effect of double ratings can might be minimized without violation of the voters privacy (but we would need the help of BonnerGuido):

1.  BonnerGuido makes an export of his GCVote database but replaces all user identifying columns (like gc user name) with a column that contains hash(`gcid of this cache` + `username` + `pw of the user`)
2. You can now import all of the votes from GCVote into you database without being able to identify the users without their help (as the `gcid` is included you are also not able to identify all caches of a user and try to identify him in that way).
3. As soon someone allows you to take over his votes von gcvote (and gives you this username and gcvote pw) you will be able to get his vote data from the gcvote-server (as you already doing it) calculate the hashes and remote the now duplicate votes from the list of anonymous votes.

Step 3 will require the GCVote database to be still active. If it should be shut down at one day it would be nice to have a db table with key value pairs of: hash(`username` + `pw of the user`) and aesEncrypt(`list of all caches the user voted for`, key=hash(`some string` + `username` + `pw of the user` )).
Every of those hash operations can and should also use salts and a good hashing algorithm of cause.

But however I do not see double voting as a real problem (we are doing avg/mean calc for the overall result) and even at gcvote I can easily create dummy accounts for manipulative votes. 

I thing the big motivation behind using GCVote was that it had such a huge userbase and a reasonable amount of votes for nearly every cache. Without that database CacheVote will have a hard time to increase it's userbase (provide votes but do not receive many).

**Comment by dev-Fabi** [comment]
@lukeIam thank you for your suggestion. For the moment I will stay with the gcVoteFallback solution, but I will definitly consider your idea and talk to BonnerGuido if this would be possible.

**Comment by dev-Fabi** [comment]
Finally I have managed to finish the API and the coresponding documentation.

It's available under https://api.cachevote.org/doc
If there are any questions or suggestions for improvemnt don't hesitate to contact me.

In future maybe an API-Token will be needed. For the moment it would be great if following header is included with each request:
**App: cgeo**

PS: cachevote has been moved to an own domain https://www.cachevote.org

**Comment by Lineflyer** [comment]
@dev-Fabi 
Thats great news!

One side question though:
While it is mentioned in this thread, that GCVote is no longer maintained is there any "official" statement regarding this from its author. The background of my question: Instead of adding CacheVote as another method of voting I would suggest to replace GCVote by CacheVote in c:geo keeping more or less the same UI features. Therefore however I would somehow like to really know, if and when GCVote will be discontinued. Up to now all functions seem to work normal. Are users still able to register for it after all?

**Comment by fm-sys** [comment]
> Are users still able to register for it after all?

I tried some months ago but wasn't able to.

**Comment by dev-Fabi** [comment]
> Instead of adding CacheVote as another method of voting I would suggest to replace GCVote by CacheVote in c:geo keeping more or less the same UI features.

I think thats the the only practical solution, because in my opinion two voting systems won't be really userfriendly...

> Therefore however I would somehow like to really know, if and when GCVote will be discontinued. Up to now all functions seem to work normal. Are users still able to register for it after all?

The API is still working but as stated on the website ([gcvote.com](https://ssl.webpack.de/gcvote.com/)) it is not maintained anymore. Also Guido said to me: "The server is running but otherwise I pretty much ignore it".
New users can't register for gcvote since quite some time now.
At the moment there is no real plan when GCVote will be discontinued. I think it really depends on "how fast" the users and services will change to CacheVote.

**Comment by eddiemuc** [comment]
To allow for graceful migration, I would suggest to add CacheVote but let the user maintain just one vote (which is then send to both systems if user has offered a working account for both).

For display, in case a cache has votes in both system, I would either display an average of it or define one of them as Superior and show that one (we might even allow the user to configure which one that is)

**Comment by Lineflyer** [comment]
> For display, in case a cache has votes in both system, I would either display an average of it or define one of them as Superior and show that one (we might even allow the user to configure which one that is)

Isn't there already data being took over from GCVote to CacheVote on server side? As I can read above I think existing votes from GCVote should already be contained in the votes coming in from CacheVote. In that case getting votes should be done from CacheVote while sending votes could be done towards both systems (if meaningful after all).

**Comment by dev-Fabi** [comment]
> Isn't there already data being took over from GCVote to CacheVote on server side?

Yes, if CacheVote itself has no votes to a cache it tries to get them from GCVote (if option is set - see API doc)

**Comment by eddiemuc** [comment]
> > Isn't there already data being took over from GCVote to CacheVote on server side?
> 
> Yes, if CacheVote itself has no votes to a cache it tries to get them from GCVote (if option is set - see API doc) 

Sorry, haven't read this before. If CacheVote is already doing the fallback work, then we don't have to do it again.

**Comment by SammysHP** [comment]
Problem is that a single vote in CacheVote overrides all votes of GCVote.

**Comment by fm-sys** [comment]
I would vote for not introducing two different voting systems but change it in cgeo. The main question would be how to do the migration from user perspective? I think we should allow new users only to register to the new CacheVote inside c:geo, but let existing users still use their old GCVote account. Or should we force them to switch? 

@dev-Fabi IMHO it is much more acceptable that votes are doubled than if a single vote overwrites everything else. Another solution would be (as you probably have not yet that many users): Store in the DB which votes are imported by the user and don't count them as long the old GCVote API is working. Do you would always sum up your votes (without imported) + GCVote votes. What do you think? 

GCVote was so popular because it had many votes. If they all get lost, this would be a major obstacle to switching to your system...

**Comment by lukeIam** [comment]
Is there a hard switch if there is any vote in CacheVote? E.g. a single
vote average in CacheVote hides 100 votes average from GCVote?
Or is there a linear fadeover (would make sense)?

**Comment by dev-Fabi** [comment]
Just changed the implementation. Now you can get a combination of both systems (with almost no dublictes).
Documentation is also updated

**Comment by fm-sys** [comment]
I played a bit around with the demos in the [API doc](https://api.cachevote.org/doc/), but it seems that when setting `includeGcVotes` it only returns the GCVotes, even if I add a new vote via the CacheVote browser addon. Could it be that there are still some changes missing at the addon/server to work properly? BTW, I didn't have a GCVotes account if that matters.

**Comment by dev-Fabi** [comment]
> Could it be that there are still some changes missing at the addon/server to work properly?

Yes the addon needs an update that this feature works properly. This update should be published within the next days. It is still in review process...

**Comment by lukeIam** [comment]
@dev-Fabi OT: any plans to make CacheVote open source?

**Comment by fm-sys** [comment]
> @dev-Fabi OT: any plans to make CacheVote open source?

This would indeed be very cool 👍 While the server side can stay closed source IMHO (since no major changes are to be expected), publishing the add on source code at Github would be very cool and maybe quite some contributes would be interested to help you...

Especially since addons can be opened as a zip archive (if you know how) so the source code is disclosed anyway, uploading it to github might bring great benefits ;-)

**Comment by fm-sys** [comment]
Thanks @dev-Fabi, I just now installed the update and everything is working fine now. You are doing really great work! Looking forward that it becomes implemented in c:geo... ;-)

**Comment by dev-Fabi** [comment]
> @dev-Fabi OT: any plans to make CacheVote open source?

I've actually been thinking about releasing the project as open source from the beginning. Especially the addon, since as @fm-sys said the source is accessible anyway.
I'm not 100% sure yet, but the tendency is towards open source. Nevertheless it will probably take some time, because I want to refactor the code and document it at least a little bit before...


> Thanks @dev-Fabi, I just now installed the update and everything is working fine now. You are doing really great work! Looking forward that it becomes implemented in c:geo... ;-)

Tank you :-)

**Comment by Lineflyer** [comment]
Someone interested in replacing GCVote by CacheVote?
I meanwhile also got some user requests on support mail with this topic and I guess that the days of GCVote are counted...

**Comment by Lineflyer** [comment]
I again heard some rumors about GCVote being shut down soon.
@dev-Fabi Do you have any insights?

@dev-Fabi I am not sure if someone from our team is able/interested/willing to work on this topic. Feel free to start working on it yourself, if you are interested?!

**Comment by dev-Fabi** [comment]
> I've heard some rumours again that GCVote will be shut down soon.
> @dev-Fabi Do you have any findings?

No, haven't heard anything new. Will check with Guido...


> @dev-Fabi Not sure if anyone on our team is able/interested/willing to work on this. Feel free to start working on it yourself if you are interested!!!

If no one else is interested, I could definitely work on it. I have a major update planned for CacheVote and right now I don't have that much time to spare. But when the update is released, I will start implementing it in c:geo.

**Comment by Lineflyer** [comment]
> But when the update is released, I will start implementing it in c:geo.

That sounds great!

**Comment by Lineflyer** [comment]
@dev-Fabi 
Just a small note in case you finished your work on the CacheVote release meanwhile:
We just merged branches to prepare a new beta version / release as well, which means that from now developers have the most time to implement and fine tune new features (like this)  on `master` branch until the next release.

**Comment by Lineflyer** [comment]
@dev-Fabi 
Just a small ping!
We are now again at the end of a release cycle as we recently released our new feature release, meaning again most time to implement a new feature.  If you had the chance to finalize your API and are interested to implement it in c:geo, you are welcome.

I guess it will help your service to get commonly known and used. :)

I would personally like to see it as c:geo is still using GCVote which can fail anytime (according to the info we have) and we thus are forced to remove it.

BTW:
We should check how c:geo behaves if the GCVote server is not responding. AFAIR we had some sticking operations in such cases back in the days.

**Comment by titanvcia** [comment]
Hello.
I have been using gcvote for years, but it has not worked for me for more than a year. At least in Spain it seems that it no longer works for anyone. Although I still see that from time to time someone, presumably a foreigner, still values ​​some of the caches here. That is why I thought that there would still be some other extension with which to vote.
I just found that cachevote replaces gcvote, I have made it public in my geocacher community, I have installed it in both Chrome and Firefox on my computer, working perfectly, but I have not achieved it in those same browsers but in android.
Also, my votes with cachevote are reflected on the web perfectly, but they are not seen in cgeo.
Is it not yet implemented in cgeo and when it is, the evaluations that I am doing will automatically be seen, or do I have to do something in cgeo to be able to see them?
Thanks.

**Comment by titanvcia** [comment]
For instance. If you look at this cache from a browser you will see that it has a 5-star rating, which I gave it with cachevote.
If you look at it from cgeo, no rating appears.
https://coord.info/GC93QVG

**Comment by rsudev** [comment]
@titanvcia The integration of cachevote is still in progress and thus not yet part of any version of c:geo (be it released, beta or nightly)

**Comment by dev-Fabi** [comment]
@Lineflyer
In the last few months I have unfortunately had little time to work on CacheVote, but I am still working on it.
As soon as I have finished the next version, the WebExtension will become OpenSource and I will also start implementing it in cgeo.

Please be patient a little longer. I will inform you here, as soon as I'm starting with the implementation.

@titanvcia 
as @rsudev said CacheVote is not implemented in cgeo yet.

**Comment by titanvcia** [comment]
Thanks for your quick responses.
And is there a way I can see cachevote ratings on the web on android?

**Comment by dev-Fabi** [comment]
> And is there a way I can see cachevote ratings on the web on android?

Yes, you should also be able to install the WebExtension on mobile firefox app (but I don't think that the geocaching.com page and the CacheVote extension are really useable on a smartphone)

**Comment by SammysHP** [comment]
@dev-Fabi 

> Yes, you should also be able to install the WebExtension on mobile firefox app (but I don't think that the geocaching.com page and the CacheVote extension are really useable on a smartphone)

Almost exactly a year ago Mozilla limited WebExtensions for Android to a handful of "recommended" extensions. Currently 16 extensions are available: https://addons.mozilla.org/en-US/android/search/?promoted=recommended&type=extension

**Comment by Lineflyer** [comment]
@dev-Fabi Any news? Are you still interested in implementing CacheVote? IMHO its again a good point in time as we can either include it in next beta (some weeks ahead) or could include it after next beta split (allowing more time to integrate fully and test).

**Comment by dev-Fabi** [comment]
@Lineflyer lately I've been working again on the new API for CacheVote, which should make "third party" integrations much easier and also allow integrations with Opencaching.
At the moment - if nothing else comes up - I have the end of February in mind for "going live" 
After that I could finally start with the implementation for c:geo.

**Comment by S-Man42** [comment]
@dev-Fabi Any news here? :) Would appreciate to see any progress on that project!

**Comment by SammysHP** [comment]
@dev-Fabi

> At the moment - if nothing else comes up - I have the end of February in mind for "going live"
After that I could finally start with the implementation for c:geo.

Two years have passed since the last update. Are you still working on CacheVote or has the project died?

**Comment by bekuno** [comment]
@dev-Fabi Any news here?

**Comment by SammysHP** [comment]
Unfortunately I have assume that CacheVote is abandoned. If @dev-Fabi decides to continue his project or transfer it to other maintainers, we can continue to work on the integration and open this issue again.

**Comment by S-Man42** [comment]
I believe in that, too. Unfortunately.

**Comment by Lineflyer** [comment]
@dev-Fabi 
As gcvote servers now finally stopped working, we will potentially remove gcvote from c:geo rather soon.
Any news on CacheVote from your side?

**Comment by ztNFny** [comment]
> @dev-Fabi 
> As gcvote servers now finally stopped working, we will potentially remove gcvote from c:geo rather soon.
> Any news on CacheVote from your side?

It s been dead for 4 years. Was never actually alive ...

**Comment by S-Man42** [comment]
Unfortunately and sadly. Yes.

**Comment by Lineflyer** [comment]
How do you judge this? I do not find any such information on the website.

At least the website and the plugin are working and I see successful communication for both retrieving and posting votes from/to the backend.

<img width="650" height="156" alt="Image" src="https://github.com/user-attachments/assets/7942ba7f-75a1-4b8e-8398-8566c10d0f38" />

**Comment by S-Man42** [comment]
Tbh, at some point I stopped watching any activity. The points you mentioned always exisited and worked fairly from the very beginning. But nonetheless they are buggy or non-working anymore on several subpages of GC.com. Any request or question over the years (e.g. the one here in the thread from 2024) are still without response, the bugs are not fixed, the website never changed (as far as I can see). So, yes, unfortunately the project seems to be not maintained anymore imho.
